ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -416,34 +335,50 @@ func (r *RayClusterReconciler) reconcilePods(ctx context.Context, instance *rayv
 	// Reconcile head Pod
 	if len(headPods.Items) == 1 {
 		headPod := headPods.Items[0]
-		r.Log.Info("reconcilePods", "head pod found", headPod.Name)
+		r.Log.Info("reconcilePods", "Found 1 head Pod", headPod.Name)
+		// TODO (kevin85421): Consider deleting a head Pod if its Ray container restarts excessively, as this
+		// might suggest an unhealthy Kubernetes node. Deleting and then recreating the head Pod might allow
+		// it to be scheduled on a different node. However, it's aggressive to delete a head Pod that is not
+		// in a terminated state (i.e., `Failed` or `Succeeded`). We should only delete a head Pod when GCS
+		// fault tolerance is enabled, and drain the head Pod before deleting it.
 		if headPod.Status.Phase == corev1.PodRunning || headPod.Status.Phase == corev1.PodPending {
-			r.Log.Info("reconcilePods", "head pod is up and running... checking workers", headPod.Name)
-		} else if headPod.Status.Phase == corev1.PodFailed && strings.Contains(headPod.Status.Reason, "Evicted") {
-			// Handle evicted pod
-			r.Log.Info("reconcilePods", "head pod has been evicted and controller needs to replace the pod", headPod.Name)
+			r.Log.Info("reconcilePods", "The head pod is Running or Pending... checking workers", headPod.Name)
+		} else {
+			if headPod.Spec.RestartPolicy == corev1.RestartPolicyAlways {
+				// Based on my observation, a Pod with `RestartPolicy: Always` will never be in the terminated states (i.e., `Failed` or `Succeeded`).
+				// However, I couldn't find any well-defined behavior in the Kubernetes documentation, so I can't guarantee that the status transition
+				// from `Running` to `Failed / Succeeded` and back to `Running` won't occur when we kill the main process (i.e., `ray start` in KubeRay)
+				// in the head Pod. Therefore, I've added this check as a safeguard.
+				message := fmt.Sprintf(
+					"The status of the head Pod %s is %s. However, KubeRay will not delete the Pod because its restartPolicy is set to 'Always' "+
+						"and it should be able to restart automatically.", headPod.Name, headPod.Status.Phase)
+				r.Log.Info(message)
+				return fmt.Errorf(message)
+			}
+			message := fmt.Sprintf("The status of the head Pod %s is %s. KubeRay will delete the Pod and recreate the head Pod in the next reconciliation.", headPod.Name, headPod.Status.Phase)
```

Comment:
The status "SUCCEEDED" for a pod indicates that all of its containers have terminated with an exit code of 0. Many applications exit with a code of 0 because they possess signal handlers that gracefully shut down the applications. However, from a deployment perspective (e.g., in Kubernetes), this might still be considered a failure.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx-pod
  labels:
    app: nginx
spec:
  containers:
  - name: nginx
    image: nginx:1.17.10
    ports:
    - containerPort: 80
  restartPolicy: Never
```

For example, create an Nginx Pod with `restartPolicy: Never`, and then kill the process PID=1 in the Pod. The Pod status will be "SUCCEEDED".

<img width="808" alt="Screen Shot 2023-08-21 at 10 56 17 AM" src="https://github.com/ray-project/kuberay/assets/20109646/4c475032-5a3f-41ab-b737-fa8a757f5885">

Reply:
Oh yeah, it makes sense to me, I was more worrying about the user who's less experienced. They might see the error message and get confused without this context